quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -515,47 +506,79 @@ runActions
+  let name = show (polarity act) ++ actionName (polarAction act)
+  monitor $ tabulate "Actions" [name]
+  monitor $ tabulate "Action polarity" [show $ polarity act]
+  case (act, ret) of
+    (ActionWithPolarity action PosPolarity, Left err) ->
+      positiveActionFailed action err
+    (ActionWithPolarity action PosPolarity, Right val) -> do
+      (s', env') <- positiveActionSucceeded action ret val
+      runSteps s' env' as
+    (ActionWithPolarity action NegPolarity, _) -> do
+      (s', env') <- negativeActionResult action ret
+      runSteps s' env' as
   where
-    loop :: Annotated state -> Env m -> [Step state] -> PropertyM m (Annotated state, Env m)
-    loop _s env [] = do
-      unless (null rejected) $
-        monitor $
-          tabulate "Actions rejected by precondition" rejected
-      return (_s, reverse env)
-    loop s env ((v := act) : as) = do
-      pre $ computePrecondition s act
-      ret <- run $ performResultToEither <$> perform (underlyingState s) (polarAction act) (lookUpVar env)
-      let name = show (polarity act) ++ actionName (polarAction act)
-      monitor $ tabulate "Actions" [name]
-      monitor $ tabulate "Action polarity" [show $ polarity act]
-      if
-        | polarity act == PosPolarity
-        , Left err <- ret -> do
-            monitor $
-              monitoringFailure @state @m
-                (underlyingState s)
-                (polarAction act)
-                (lookUpVar env)
-                err
-            stop False
-        | otherwise -> do
-            let var = unsafeCoerceVar v
-                s' = computeNextState s act var
-                env'
-                  | Right val <- ret = (var :== val) : env
-                  | otherwise = env
-            monitor $ monitoring @state @m (underlyingState s, underlyingState s') (polarAction act) (lookUpVar env') ret
-            (b, (Endo mon, Endo onFail)) <-
-              run
-                . runWriterT
-                . runPost
-                $ computePostcondition @m
-                  (underlyingState s, underlyingState s')
-                  act
-                  (lookUpVar env)
-                  ret
-            monitor mon
-            unless b $ monitor onFail
-            assert b
-            loop s' env' as
+    positiveActionFailed action err = do
+      monitor $
+        monitoringFailure @state @m
+          (underlyingState s)
+          action
+          (lookUpVar env)
+          err
+      stop False
+
+    positiveActionSucceeded action ret val =
+      computeActionResult action ret $ \stateTransition ->
+        postcondition
+          stateTransition
+          action
+          (lookUpVar env)
+          val
+
+    negativeActionResult action ret =
+      computeActionResult action ret $ \stateTransition ->
+        postconditionOnFailure
+          stateTransition
+          action
+          (lookUpVar env)
+          ret
```

Comment:
These are very nice, they make things clearer. I'm very torn on my above comment. The other option here is to get rid of `computeNextState` too and inline things _more_ here instead.

Reply:
You mean get rid of `computeActionResult`? That could be clearer indeed, let me try.

Reply:
Possibly both! `computeNextState` cases on the polarity in a way that might be unnecessary given this factoring.

Reply:
On the other hand, it does allow an awful lot of _other_ code to be cleaner than it would be otherwise.